File: claude-controller/state.py
"""Persisted agent-wide state across runs.

Single JSON object on disk with per-feature top-level keys. New keys can be
added without migrations; missing keys read back as absent. Today the only
key is ``sectool_version_check`` (the staleness-check cache).
"""

from __future__ import annotations

import json
import os
import tempfile


def load(path: str) -> dict:
    """Return the dict at path; missing or corrupt files yield ``{}``."""
    try:
        with open(path) as f:
            data = json.load(f)
    except (FileNotFoundError, OSError, ValueError):
        return {}
    if not isinstance(data, dict):
        return {}
    return data


def save(path: str, state: dict) -> None:
    """Write state to path atomically via tmp + rename."""
    directory = os.path.dirname(path) or "."
    fd, tmp_path = tempfile.mkstemp(
        prefix=os.path.basename(path) + ".tmp-",
        dir=directory,
    )
    try:
        with os.fdopen(fd, "w") as f:
            json.dump(state, f)
        os.replace(tmp_path, path)
    except OSError:
        try:
            os.remove(tmp_path)
        except OSError:
            pass

File: secagent/state/state.go
// Package state persists agent-wide state across runs.
//
// The on-disk schema is a single JSON object whose top-level fields are
// per-feature sub-structs with `omitempty` JSON tags, so adding a new
// kind of state is a one-field change. Today the only field is the
// best-effort sectool version-check cache.
package state

import (
	"encoding/json"
	"errors"
	"fmt"
	"os"
	"path/filepath"
)

// State is the full on-disk shape. New per-feature fields go here.
type State struct {
	SectoolVersionCheck *SectoolVersionCheckEntry `json:"sectool_version_check,omitempty"`
}

// SectoolVersionCheckEntry caches the staleness-check outcome.
// Status is "ok" or "failed"; LatestVersion is only set when Status == "ok".
type SectoolVersionCheckEntry struct {
	LatestVersion string `json:"latest_version,omitempty"`
	CheckedAt     int64  `json:"checked_at"`
	Status        string `json:"status"`
}

// Load returns the state at path; missing or corrupt files yield an empty State without error.
func Load(path string) (*State, error) {
	data, err := os.ReadFile(path)
	if err != nil {
		if errors.Is(err, os.ErrNotExist) {
			return &State{}, nil
		}
		return nil, fmt.Errorf("read state: %w", err)
	}
	var s State
	if err := json.Unmarshal(data, &s); err != nil {
		// corrupt -> reset rather than propagate
		return &State{}, nil
	}
	return &s, nil
}

// Save writes s to path atomically via tmp+rename.
func Save(path string, s *State) error {
	data, err := json.Marshal(s)
	if err != nil {
		return fmt.Errorf("marshal state: %w", err)
	}
	dir := filepath.Dir(path)
	tmp, err := os.CreateTemp(dir, filepath.Base(path)+".tmp-*")
	if err != nil {
		return fmt.Errorf("create tmp: %w", err)
	}
	tmpPath := tmp.Name()
	if _, err := tmp.Write(data); err != nil {
		_ = tmp.Close()
		_ = os.Remove(tmpPath)
		return fmt.Errorf("write tmp: %w", err)
	}
	if err := tmp.Close(); err != nil {
		_ = os.Remove(tmpPath)
		return fmt.Errorf("close tmp: %w", err)
	}
	if err := os.Rename(tmpPath, path); err != nil {
		_ = os.Remove(tmpPath)
		return fmt.Errorf("rename: %w", err)
	}
	return nil
}
